### PR TITLE
Reference residual improvements

### LINKED ISSUE #18772: Update ReferenceResidual problem
## Reason
Remove dead code, clean up, add option to use only some of the solution variables for convergence checking. Add globbing to allow specifying variable names with wildcards.

## Design
Add globbing helper in MooseUtils.

## Impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #18772

### CODE REVIEW

File: framework/src/problems/ReferenceResidualProblem.C [hunk 1]
```diff
@@ -44,23 +44,29 @@ ReferenceResidualProblem::validParams()
   params.addParam<int>("acceptable_iterations",
                        0,
                        "Iterations after which convergence to acceptable limits is accepted");
-  params.addParam<std::vector<std::vector<std::string>>>(
+  params.addParam<std::vector<std::vector<NonlinearVariableName>>>(
       "group_variables",
       "Name of variables that are grouped together to check convergence. (Multiple groups can be "
       "provided, separated by semicolon)");
+  params.addParam<std::vector<NonlinearVariableName>>(
+      "converge_on",
```
Comment (tophmatthews) [inline framework/src/problems/ReferenceResidualProblem.C:52]: 🥳

**Comment by dschwen** [comment]
@bwspenc Please. Review. My. Pull. Requests.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: framework/include/problems/ReferenceResidualProblem.h [hunk 2]
```diff
@@ -64,37 +62,27 @@ class ReferenceResidualProblem : public FEProblem
...
-   * @param ref_vars A set of reference residual variables that need to be added to
-   * ReferenceResidualProblem.
-   */
-  void addReferenceResidualVariables(std::set<std::string> & ref_vars);
-
-  /**
-   * Add a set of variables that need to be grouped together.
+   * Add a set of variables that need to be grouped together. For use in
+   * actions that create variables. This is templated for backwards compatibility as
+   * Bison uses std::string where we prefer to use NonlinearVariableName.
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/include/problems/ReferenceResidualProblem.h:67]: disd't mean to code shame Bison...

File: framework/src/problems/ReferenceResidualProblem.C [hunk 3]
```diff
@@ -361,35 +328,40 @@ ReferenceResidualProblem::updateReferenceResidual()
...
   }
 
   for (unsigned int i = 0; i < _soln_vars.size(); ++i)
   {
-    _resid[i] = s.calculate_norm(nonlinear_sys.RHS(), _soln_vars[i], DISCRETE_L2);
-    _group_resid[_variable_group_num_index[i]] += Utility::pow<2>(_resid[i]);
+    const auto resid =
+        Utility::pow<2>(s.calculate_norm(nonlinear_sys.RHS(), _soln_vars[i], DISCRETE_L2));
+    const auto group = _variable_group_num_index[i];
+    _group_resid[group] += _converge_on_var[i] ? resid : 0;
```
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/problems/ReferenceResidualProblem.C:340]: That check isn't quite so easy, as all variables are in groups (some are just in groups of size 1)

[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/problems/ReferenceResidualProblem.C:340]: actually it is easy..

[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/problems/ReferenceResidualProblem.C:347]: done

[the remark this answers is not in the document]
Comment (bwspenc) [reply framework/src/problems/ReferenceResidualProblem.C:347]: I think we wouldn't ever fall into the 0 case on line 347 now, would we? Seems like a good use for an assert.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Minor issue with requirement being wrong. Also see my note about a potential use of an assert.

File: test/tests/problems/reference_residual_problem/tests [hunk 4]
```diff
@@ -23,4 +23,25 @@
...
+    expect_out = 'Solve Converged!'
+    cli_args = 'Problem/converge_on="u"'
+    requirement = 'The system shall have the ability to base system convergence on the convergence of a subset of variables.'
+  []
+  [converge_on_group_error]
+    type = 'RunException'
+    input = 'reference_residual.i'
+    expect_err = 'You added variable \'u\' to a group but excluded it from the convergence check. This is not permitted.'
+    cli_args = 'Problem/converge_on="v" Problem/group_variables="u v"'
+    requirement = 'The system shall have the ability to base system convergence on the convergence of a subset of variables.'
```
[the remark this answers is not in the document]
Comment (dschwen) [reply test/tests/problems/reference_residual_problem/tests:38]: D'Uh, right, forgot to change that

File: framework/src/problems/ReferenceResidualProblem.C (hunk 3 above)
[the remark this answers is not in the document]
Comment (dschwen) [reply framework/src/problems/ReferenceResidualProblem.C:347]: we absolutely will fall into that regularly when variables are excluded from the check

[the remark this answers is not in the document]
Comment (bwspenc) [reply framework/src/problems/ReferenceResidualProblem.C:347]: Oh yeah, because everything's grouped...

**Review by bwspenc — APPROVED** [review]

**Pull request merged**